### ISSUE TITLE
scale loss per token/local sequence for discrete system representation

### ISSUE THREAD

**Issue by wesboyt** [huggingface/transformers#38854]
### Feature request

I have implemented a scaling loss for my discrete game usecase.
I would like to be able to hand my model a batch with a map of sequences and their associated randomness(or importance) and then scale the loss relative to the next tokens predictability.
My implementation works but is very slow relative to: 
https://github.com/huggingface/transformers/blob/v4.52.3/src/transformers/trainer.py#L3795

### Motivation

My system has highly variable regions that are truly random and prevent my models from converging without manual training which is incredibly inefficient.

### Your contribution

My implementation:
            lookup = torch.zeros(86,).to(self.device)
            lookup[self.x_token] = 2
            lookup[self.y_token] = 2
            lookup[self.z_token] = 1
            loss_tracker = torch.zeros(batch_size, ).to(self.device)
            for i in full_range:
                label = input_padded[:,i]
                logits = self.model(input_padded[:,:i]).logits[:,-1]
                pre = i <= 26
                if pre:
                    modifier = torch.full((batch_size,), .05, dtype=torch.float32).to(self.device)
                    loss = (loss_function(logits, label) * modifier).mean()
                else:
                    modifier = torch.full((batch_size,), .1, dtype=torch.float32).to(self.device)
                    modifier[loss_tracker > 0] = 1.0
                    loss = (loss_function(logits, label) * modifier).mean()
                loss_tracker -= 1.0
                special_indexes = torch.argwhere(label == special_ids)
                if i < max_length - 1:
                    next = input_padded[:,i+1]
                    next_lookup = hero_lookup[next]
                    loss_tracker[special_indexes] += next_lookup[special_indexes]

Basically if its the begining of a sequence there is low information and extreme randomness, I lower my loss by 95%.
Otherwise I use a loss tracker to denote if a subsequence is predictable and for how many tokens I should use full loss.
I adjust less important sequences to 10% loss and keep important sequences at full loss.
This impl supports batching but is highly specific to my discrete system which I've turned into a onehot language.
I believe there are many other use cases where identifying a token or sequence of tokens and altering the loss thereafter when the user understands the physics of the system being simulated is extremely useful.
I improved accuracy for these desireable and rarer sequences from 88% to 96% in only 20000 sentences.

**Comment by Rocketknight1** [comment]
Hi @wesboyt, scaling classification loss with the sequence entropy is a cool idea and could definitely be relevant in some domains, including DNA. However, it's quite advanced and extremely domain-specific, so I don't think we should add it to the `Trainer`! cc @sunmarc to confirm.

**Comment by wesboyt** [comment]
Was just comparing the performance to trainer loop (which is roughly 15x faster).
Would not intend to tell your team how or where to implement it.

**Comment by wesboyt** [comment]
I could imagine a impl where the user just hands another field with the batch, loss_modifier or something. with the expectation being they iterated over the inputs and calculated which tokens deserved which loss modifier.
batch{
    inputs:[[1,2,3]],
    loss_mod:[[0.05,0.1,1.0]],
    attention:[[1,1,1]]
}
